paas_app_charmer/go/charm.py

```diff
@@ -0,0 +1,89 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Go Charm service."""
+
+import pathlib
+import typing
+
+import ops
+from pydantic import BaseModel, Extra, Field
+
+from paas_app_charmer.app import App, WorkloadConfig
+from paas_app_charmer.charm import PaasCharm
+
+
+class GoConfig(BaseModel, extra=Extra.allow):
+    """Represent Go builtin configuration values.
+
+    Attrs:
+        port: port where the application is listening
+        metrics_port: port where the metrics are collected
+        metrics_path: path where the metrics are collected
+        secret_key: a secret key that will be used for securely signing the session cookie
+            and can be used for any other security related needs by your Flask application.
+    """
+
+    port: int = Field(alias="port", default=8080, gt=0)
+    metrics_port: int | None = Field(alias="metrics-port", default=None, gt=0)
+    metrics_path: str | None = Field(alias="metrics-path", default=None, min_length=1)
+    secret_key: str | None = Field(alias="secret-key", default=None, min_length=1)
```

Comment:
One thing is not clear to me from this PR: how will we be able to pass arbitrarty configs declared in `charmcraft.yaml` to the workload/Go application?

Reply:
I have added (also related to your previous comment) a new integration test with a config in `charmcraft.yaml`. There is no need to change code (except in charmcraft.yaml) to add a new environment variable. The pydantic structure is related to the "framework" configuration, not user defined configurations.

Reply:
Nice! Say a config is required by the application. How/where should the `Charm` class perform this kind of validation and set the status to blocked?

Reply:
Unfortunately, for a configuration option added by a user, the `charmcraft.yaml` file format has no option to set it as optional/mandatory or to specify constraints. This is different from integrations, as they can be optional or mandatory.

As it is now, and until there is a better solution, the user may for example override [is_ready](https://github.com/canonical/paas-app-charmer/blob/b9bbe4766de0d88a1b920e3360a9a9b10ba63fef/paas_app_charmer/charm.py#L218), and return `False` if there is something wrong (the base is_ready must also be called or the library will not work correctly).

Reply:
I will add it to our documentation stories so we can discuss about it in the team. Thanks for the advice!